Cabal/src/Distribution/Simple/Utils.hs

```diff
@@ -372,7 +379,61 @@ die' verbosity msg = withFrozenCallStack $ do
     =<< annotateErrorString verbosity
     =<< pure . wrapTextVerbosity verbosity
     =<< pure . addErrorPrefix
-    =<< prefixWithProgName msg
+    =<< prefixWithProgName msg      
+
+-- Type which will be a wrapper for cabal -expections and cabal-install exceptions
+data VerboseException a  = VerboseException CallStack POSIXTime Verbosity a
+ deriving (Show, Typeable)
+
+-- A new dieWithException function which will replace the existing die' call sites
+dieWithException :: HasCallStack => Verbosity -> CabalException -> IO a
+dieWithException verbosity exception = do
+  ts <- getPOSIXTime
+  case exception of
+    NoBenchMarkProgram path -> throwIO $ VerboseException callStack ts verbosity (NoBenchMarkProgram path)
+    EnableBenchMark -> throwIO $ VerboseException callStack ts verbosity EnableBenchMark
+    BenchMarkNameDisable bmName -> throwIO $ VerboseException callStack ts verbosity (BenchMarkNameDisable bmName)
+    NoBenchMark bmName -> throwIO $ VerboseException callStack ts verbosity (NoBenchMark bmName)
+    NoLibraryFound   -> throwIO $ VerboseException callStack ts verbosity NoLibraryFound
+    CompilerNotInstalled compilerFlavor -> throwIO $ VerboseException callStack ts verbosity (CompilerNotInstalled compilerFlavor)
+    CantFindIncludeFile file -> throwIO $ VerboseException callStack ts verbosity (CantFindIncludeFile file)
+    SourceDistException errormsg -> throwIO $ VerboseException callStack ts verbosity (SourceDistException errormsg) 
+
+instance (Show a ,Typeable a) => Exception (VerboseException a) where
+_displayException :: VerboseException CabalException -> [Char]
+
```

Comment:
Similarly all these cases can be reduced away to a single case -- no need to pattern match on each individual possibility -- they're all treated uniformly.

Reply:
Incorporated the comment. Thanks